usage/src/org/apache/cloudstack/quota/QuotaManagerImpl.java

```diff
@@ -0,0 +1,465 @@
+        QuotaTariffVO tariff = _quotaTariffDao.findTariffPlanByUsageType(transferType, usageRecord.getEndDate());
+        if (tariff != null && !tariff.getCurrencyValue().equals(0)) {
+            BigDecimal onegbcost;
+            BigDecimal rawusageingb;
+            BigDecimal networkusage;
+            onegbcost = tariff.getCurrencyValue();
+            // s_logger.info("Quotatariff onegbcost=" + onegbcost);
+            rawusageingb = new BigDecimal(usageRecord.getRawUsage()).divide(s_gb, 8, RoundingMode.HALF_EVEN);
+            networkusage = rawusageingb.multiply(onegbcost);
+            quota_usage = new QuotaUsageVO(usageRecord.getId(), usageRecord.getZoneId(), usageRecord.getAccountId(), usageRecord.getDomainId(), transferType, networkusage, usageRecord.getStartDate(),
+                    usageRecord.getEndDate());
+            _quotaUsageDao.persist(quota_usage);
+        }
+
+        usageRecord.setQuotaCalculated(1);
+        _usageDao.persist(usageRecord);
+        return quota_usage;
+    }
+
+    public Date startOfNextDay() {
+        Calendar c = Calendar.getInstance();
+        c.setTime(new Date());
+        c.add(Calendar.DATE, 1);
+        Date dt = c.getTime();
+        return dt;
+    }
+
+    protected boolean lockAccount(long accountId) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy.open(TransactionLegacy.CLOUD_DB).close();
+        boolean success = false;
+        Account account = _accountDao.findById(accountId);
+        if (account != null) {
+            if (account.getState().equals(State.locked)) {
+                return true; // already locked, no-op
+            } else if (account.getState().equals(State.enabled)) {
+                AccountVO acctForUpdate = _accountDao.createForUpdate();
+                acctForUpdate.setState(State.locked);
+                success = _accountDao.update(Long.valueOf(accountId), acctForUpdate);
+            } else {
+                if (s_logger.isInfoEnabled()) {
+                    s_logger.info("Attempting to lock a non-enabled account, current state is " + account.getState() + " (accountId: " + accountId + "), locking failed.");
+                }
+            }
+        } else {
+            s_logger.warn("Failed to lock account " + accountId + ", account not found.");
+        }
+        TransactionLegacy.open(opendb).close();
+        return success;
+    }
+
+    public boolean enableAccount(long accountId) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy.open(TransactionLegacy.CLOUD_DB).close();
+        boolean success = false;
+        AccountVO acctForUpdate = _accountDao.createForUpdate();
+        acctForUpdate.setState(State.enabled);
+        acctForUpdate.setNeedsCleanup(false);
+        success = _accountDao.update(Long.valueOf(accountId), acctForUpdate);
+        TransactionLegacy.open(opendb).close();
```

Comment:
LegacyTranscaction is in use in cloud almost everywhere. While this is not a reason to continue its use. My thinking is that this should be delinked from Quota changes and looked at in the right perspective.
We should make a joint effort to clear this up.